server/src/main/java/org/elasticsearch/action/search/AbstractSearchAsyncAction.java

```diff
@@ -236,6 +228,10 @@ public final void run() {
             assert iterator.skip();
             skipShard(iterator);
         }
+        final Map<SearchShardIterator, Integer> shardIndexMap = Maps.newHashMapWithExpectedSize(shardIterators.length);
+        for (int i = 0; i < shardIterators.length; i++) {
+            shardIndexMap.put(shardIterators[i], i);
+        }
```

Comment:
This is fine because run is always every called once on a specific async action instance?

Reply:
Yea, the results array among other things is pretty much single-use :)